schematics/exceptions.py

```diff
@@ -2,12 +2,76 @@
 
 from __future__ import unicode_literals, absolute_import
 
+import copy
+
 from collections import Sequence
 
 from .common import *
 
 
 @str_compat
+class BaseError(Exception):
+
+    def __init__(self, message, errors=None, *args):
+        """
+        The base class for all Schematics errors.
+
+        message should be a human-readable message,
+        while errors is a machine-readable list, or dictionary.
+
+        if None is passed as the message, and error is populated,
+        the primitive representation will be serialized.
+
+        the Python logging module expects exceptions to be hashable
+        and therefore immutable. As a result, it is not possible to
+        mutate BaseError's error list or dict after initialization.
+        """
+        if message is None and errors is not None:
+            message = str(self._to_primitive(errors))
+        super(BaseError, self).__init__(message)
+        self._errors = errors
+
+    @property
+    def errors(self):
+        return copy.deepcopy(self._errors)
```

Comment:
I'd vote for frozendict, the dict representation is nice without having to serialize to / from tuples. 

Would you like me to pull frozendict as a dependency, or write one for schematics?